tracer/src/Datadog.Trace.ClrProfiler.Native/always_on_profiler.cpp

```diff
@@ -284,41 +284,58 @@ class SamplingHelper
     }
 
 private:
-    void GetFunctionName(FunctionID funcID, const COR_PRF_FRAME_INFO frameInfo, shared::WSTRING& result)
+    [[nodiscard]] FunctionIdentifier GetFunctionIdentifier(const FunctionID func_id, const COR_PRF_FRAME_INFO frame_info) const
     {
-        constexpr auto unknown_list_of_arguments = WStr("(unknown)");
-        constexpr auto unknown_function_name = WStr("Unknown(unknown)");
-        constexpr auto name_separator = WStr(".");
-
-        if (funcID == 0)
+        if (func_id == 0)
         {
-            constexpr auto unknown_native_function_name = WStr("Unknown_Native_Function(unknown)");
-            result.append(unknown_native_function_name);
-            return;
+            constexpr auto zero_valid_function_identifier = FunctionIdentifier{0, 0, true};
+            return zero_valid_function_identifier;
         }
 
         ModuleID moduleId = 0;
         mdToken functionToken = 0;
-
         // theoretically there is a possibility to use GetFunctionInfo method, but it does not support generic methods
-        HRESULT hr = info10->GetFunctionInfo2(funcID, frameInfo, nullptr, &moduleId, &functionToken, 0, nullptr, nullptr);
+        HRESULT hr = info10->GetFunctionInfo2(func_id, frame_info, nullptr, &moduleId, &functionToken, 0, nullptr, nullptr);
         if (FAILED(hr))
         {
             Logger::Debug("GetFunctionInfo2 failed. HRESULT=0x", std::setfill('0'), std::setw(8), std::hex, hr);
+            constexpr auto zero_invalid_function_identifier = FunctionIdentifier{0, 0, false};
+            return zero_invalid_function_identifier;
+        }
+
+        return FunctionIdentifier{functionToken, moduleId, true};
+    }
+
+    void GetFunctionName(FunctionIdentifier functionIdentifier, shared::WSTRING& result)
+    {
+        constexpr auto unknown_list_of_arguments = WStr("(unknown)");
+        constexpr auto unknown_function_name = WStr("Unknown(unknown)");
+        constexpr auto name_separator = WStr(".");
+
+        if (!functionIdentifier.is_valid)
+        {
             result.append(unknown_function_name);
             return;
         }
 
+        if (functionIdentifier.function_token == 0)
+        {
+            constexpr auto unknown_native_function_name = WStr("Unknown_Native_Function(unknown)");
```

Comment:
nit: I think here it could be dynamic code, until we are certain it is incorrect to say native.
```suggestion
            constexpr auto unknown_native_function_name = WStr("Unknown_Metadata_Token(unknown)");
```

Reply:
For now, I will keep as it is. I have checked 
```
    internal static class ClassENonStandardCharactersĄĘÓŁŻŹĆąęółżźśćĜЖᏳⳄʤǋₓڿଟഐቐ〣‿؁੮ᾭ_<TClass>
    {
        public static void GenericMethodDFromGenericClass<TMethod, TMethod2>(TClass classArg, TMethod methodArg, TMethod2 additionalArg)
        {
            dynamic test = new Test();

            test();

        }
    }

    internal class Test : DynamicObject
    {

        public override bool TryInvoke(InvokeBinder binder, object[] args, out object result)
        {
            ClassVb.MethodVb("testParam");
            result = null;
            return true;
        }
    }
```

and it produces `dev-otel-collector-1  |         at My.Custom.Test.Namespace.Test.TryInvoke(System.Dynamic.InvokeBinder, System.Object[], System.Object&)`

Also `dev-otel-collector-1  |         at Unknown_Native_Function(unknown)` is always produces as the root of the stack trace. Please let me know what is scenario you wolud like to cover. I will try to do it or create separate story.

Reply:
This is fine, we will dig into later.